src/main/utility/mod.rs

```diff
@@ -189,27 +191,34 @@ impl Clone for ObjectCounter {
     }
 }
 
-pub fn tilde_expansion(path: &str) -> std::path::PathBuf {
+/// Resolves the path (but doesn't canonicalize)
+/// "/path/to/test" -> "/path/to/test"
+/// "~/test" -> "/home/user/test"
+/// "../test" -> "../test"
+/// "./test" -> "./test"
+/// "test" -> absolute path of test in $PATH
+pub fn resolve_path(path: &Path) -> anyhow::Result<PathBuf> {
     // if the path begins with a "~"
-    if let Some(x) = path.strip_prefix('~') {
-        // get the tilde-prefix (everything before the first separator)
-        let (tilde_prefix, remainder) = x.split_once('/').unwrap_or((x, ""));
-
-        if tilde_prefix.is_empty() {
-            if let Ok(ref home) = std::env::var("HOME") {
-                return [home, remainder].iter().collect::<std::path::PathBuf>();
-            }
-        } else if ['+', '-'].contains(&tilde_prefix.chars().next().unwrap()) {
-            // not supported
+    if let Ok(remainder) = path.strip_prefix("~") {
+        if let Ok(ref home) = std::env::var("HOME") {
+            let mut path = PathBuf::with_capacity(home.len() + remainder.as_os_str().len());
+            path.push(home);
+            path.push(remainder);
+            Ok(path)
+        } else {
+            anyhow::bail!("Couldn't find $HOME in env")
+        }
+    } else {
+        // Source exec path from $PATH if path is not a path
+        let path_str = path.to_str().ok_or(anyhow::anyhow!("invalid unicode"))?;
+        if !path_str.is_empty() && path_str.chars().find(|c| *c == '/' || *c == '.').is_none() {
+            which::which(path)
+                .with_context(|| format!("Failed to find path in $PATH for: {:?}", path))
```

Comment:
Does this work correctly if the binary name has a '.' character in it? For example `foo.a`?

Reply:
Hmm... forgot this in the newest push, will add this now.

Reply:
wait... can shell scripts be sourced from path?

Reply:
> wait... can shell scripts be sourced from `$PATH`?

Yep as long as they're executable and have a shebang header, they're not very useful in Shadow though since we don't support fork/exec. But you could have non-shell files with a '.' in PATH as well (for example on my system I have a `/bin/python3.8`).

Reply:
alr i'll fix that :+1: